solver/src/settlement_submission/archer_settlement.rs

```diff
@@ -0,0 +1,377 @@
+// manually fund it like we do with other solver accounts and make part of our monitor scripts.
+
+// TODO: Node gas estimates seem a little higher than actual. This can be seen when comparing
+// to tenderly simulations and mined transactions. This causes us to overpay somewhat.
+
+use super::{archer_api::ArcherApi, ESTIMATE_GAS_LIMIT_FACTOR};
+use crate::{interactions::block_coinbase, settlement::Settlement};
+use anyhow::{anyhow, Context, Result};
+use chrono::{DateTime, Utc};
+use contracts::GPv2Settlement;
+use ethcontract::{errors::MethodError, transaction::Transaction, Account, GasPrice};
+use futures::FutureExt;
+use gas_estimation::GasPriceEstimating;
+use primitive_types::{H256, U256};
+use shared::Web3;
+use std::{
+    sync::Arc,
+    time::{Duration, Instant, SystemTime},
+};
+use web3::types::TransactionId;
+
+pub struct ArcherSolutionSubmitter {
+    pub web3: Web3,
+    pub contract: GPv2Settlement,
+    pub account: Account,
+    pub archer_api: ArcherApi,
+    pub gas_price_estimator: Arc<dyn GasPriceEstimating>,
+    pub gas_price_cap: f64,
+}
+
+impl ArcherSolutionSubmitter {
+    /// Submit a settlement to the contract, updating the transaction with gas prices if they increase.
+    ///
+    /// Goes through the archerdao network so that failing transactions do not get mined and thus do
+    /// not cost gas.
+    ///
+    /// Returns None if the deadline is reached without a mined transaction.
+    ///
+    /// Only works on mainnet.
+    pub async fn submit(
+        &self,
+        target_confirm_time: Duration,
+        deadline: SystemTime,
+        settlement: Settlement,
+        gas_estimate: U256,
+    ) -> Result<Option<H256>> {
+        let nonce = self.nonce().await?;
+
+        tracing::info!(
+            "starting archer solution submission at nonce {} with deadline {}",
+            nonce,
+            DateTime::<Utc>::from(deadline).to_rfc3339(),
+        );
+
+        let mut transactions = Vec::new();
+        let submit_future = self.submit_with_increasing_gas_prices_until_simulation_fails(
+            deadline,
+            target_confirm_time,
+            nonce,
+            settlement,
+            gas_estimate,
+            &mut transactions,
+        );
+
+        let nonce_future = self.wait_for_nonce_to_change(nonce);
+
+        let deadline_future = tokio::time::sleep(
+            deadline
+                .duration_since(SystemTime::now())
+                .unwrap_or_else(|_| Duration::from_secs(0)),
+        );
+
+        futures::select! {
+            method_error = submit_future.fuse() => tracing::info!("stopping submission because simulation failed: {:?}", method_error),
+            new_nonce = nonce_future.fuse() => tracing::info!("stopping submission because account nonce changed to {}", new_nonce),
+            _ = deadline_future.fuse() => tracing::info!("stopping submission because deadline has been reached"),
+        };
```

Comment:
`futures` rocks! I find this quite easy to follow, and like the use of `select` as a way of "racing" to see which one finished first.